lib/Sema/CSGen.cpp

```diff
@@ -1723,9 +1723,15 @@ namespace {
     }
 
     Type visitTypeValueExpr(TypeValueExpr *E) {
-      auto ty = E->getParamDecl()->getDeclaredInterfaceType();
-      auto paramType = CS.DC->mapTypeIntoContext(ty);
-      E->setParamType(paramType);
+      auto declRefRepr = cast<DeclRefTypeRepr>(E->getRepr());
+      auto resolvedTy = resolveTypeReferenceInExpression(declRefRepr,
+                                                         TypeResolverContext::InExpression,
+                                                         CS.getConstraintLocator(E));
+
+      if (!resolvedTy || resolvedTy->hasError())
+        return Type();
+
+      E->setParamType(resolvedTy);
```

Comment:
I don't understand what is going on here because the type that is being resolved is not the type being returned and `getParamType()` not used anywhere in the solver as far as I can see... Why are we doing it here and not in PreCheckTarget somewhere if it exists for validation purposes?

Reply:
Can this access use UnresolvedDotExpr and a special OverloadChoiceKind and indicate that we found a value generic reference and only create TypeValueExpr during CSApply?

Reply:
Than we need to start keeping track of that but I think a better option would be not to create this expression at all until CSApply.

Reply:
We already generate UnresolvedDotExprs for things like `A<123>.n` and create the TypeValueExpr in CSApply. CSGen is only handling references like `n` where we have no base to attach to create an UnresolvedDotExpr. It occurs in the same place where we eagerly make TypeExpr and just check that if we are trying to make a TypeExpr for a value generic just generate a TypeValurExpr.

Reply:
I was thinking DeclRefExpr to "self" + UnresolvedDotExpr like we do for members?

Reply:
Would that work in a static context?

```swift
struct A<let n: Int> {
  static var thing: Int {
    n
  }
}
```

Reply:
This won't work in function contexts however:
```swift
func hello<let n: Int>(...) -> Int {
  n
}
```
What is `n` supposed to be a dot reference to?

Reply:
Can we still use TypeExpr for that with tailored CSGen?  I'm fine with using TypeValueExpr as a type-checked form but I'm not certain that we cannot model unresolved form with what we already have.